modules/core/src/smithy4s/Validator.scala

```diff
@@ -0,0 +1,61 @@
+/*
+ *  Copyright 2021-2024 Disney Streaming
+ *
+ *  Licensed under the Tomorrow Open Source Technology License, Version 1.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *     https://disneystreaming.github.io/TOST-1.0.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package smithy4s
+
+trait Validator[A, B] {
+  def validate(value: A): Either[String, B]
+}
+
+object Validator {
+
+  def of[A, B]: PartialValidatorBuilder[A, B] =
+    new PartialValidatorBuilder[A, B]
+
+  private[smithy4s] class PartialValidatorBuilder[A, B] {
+    def validating[C](constraint: C)(implicit
+        bijection: Bijection[A, B],
+        ev: RefinementProvider.Simple[C, A]
+    ): ValidatorBuilder[A, B] =
+      new ValidatorBuilder[A, B](ev.make(constraint), bijection)
+  }
+
+  private[smithy4s] class ValidatorBuilder[A, B] private (
+      first: Refinement.Aux[_, A, A],
+      other: List[Refinement.Aux[_, A, A]],
+      bijection: Bijection[A, B]
+  ) {
+
+    def this(ref: Refinement.Aux[_, A, A], bijection: Bijection[A, B]) =
+      this(ref, List.empty, bijection)
```

Comment:
**nitpick:** I'd replace this auxiliary constructor with an explicit parameter in the use site (it's just the one in `PartialValidatorBuilder`, right?